src/control/server/storage/provider.go

```diff
@@ -666,6 +678,33 @@ func (p *Provider) UpdateBdevFirmware(req NVMeFirmwareUpdateRequest) (*NVMeFirmw
 	return p.bdev.UpdateFirmware(req)
 }
 
+// ValidateBdevConfig updates an existing SPDK bdev config, if necessary.
+func (p *Provider) ValidateBdevConfig(ctx context.Context, ctrlrs NvmeControllers) error {
+	if !p.HasBlockDevices() {
+		return nil
+	}
+
+	_, err := p.ReadNvmeConfig(ctx)
+	if err == nil {
+		// For now, we'll just assume that if we can read the config file, then
+		// we don't need to do anything else.
+		return nil
+	}
+
+	// Take the conservative approach that if we expect there to have been a config
+	// file and it's not there, then we need the admin to investigate.
+	if !fault.IsFaultCode(err, code.SpdkInvalidConfiguration) {
+		p.log.Errorf("Failed to read bdev config file: %s", p.engineStorage.ConfigOutputPath)
+		return err
+	}
+
+	// Otherwise, if the config file is there but we can't understand it, then
+	// it's probably from a different version of DAOS and we should just regenerate
+	// it based on our version of the configuration.
+	p.log.Notice("The bdev config file was unparsable; regenerating it.")
+	return p.WriteNvmeConfig(ctx, p.log, ctrlrs)
```

Comment:
I'm unsure as to whether this is a safe approach, it seems like this scenario indicates something is wrong and maybe it is better to fail here. Format should be where all the checks and balances are accounted for, allowing ad hoc rewriting of config file seems risky. @kjacque thoughts?

Reply:
not necessarily, I'm more suggesting that rewriting the config might be hiding significant problems

Reply:
in what situation would we want to rewrite the config, if metadata directory contains corrupt files don't we want to refuse to start?

Reply:
Thanks for the extra context. In terms of upgrade path from 2.6->2.8 if we are supporting storage compatibility without reformat then this does sound reasonable as an approach.

Reply:
I think there is some value in checking whether the file is valid for spdk v22 before upgrading it. I.e. detecting the difference between an old file and a corrupt file. To be honest I think it is always safe to replace, but if we detect corruption, that would be potentially useful to log, to alert the admin.

Reply:
Hmm. As I looked into this, I found that there didn't seem to be a straightforward way to make this happen. There are layering concerns with either:
  * Pulling `bdev.SpdkConfig` and friends up into the storage layer so that we can get rid of the intermediate representation in `storage.BdevWriteConfigRequest` (and also more easily return the parsed configuration from `spdkConfig.ReadConfig()`)
  * Pulling the config validation logic down from `storage.Provider.UpgradeBdevConfig()`

Encapsulating all of the spdk config stuff in the `bdev` layer does seem to make sense. I was thinking that we might want more control over it in daos_server vs daos_server_helper, but maybe it doesn't matter.